### ISSUE TITLE
GradNormFunction: Clarification on Normalization Dimension

### ISSUE THREAD

**Issue by jaivardhankapoor** [cloneofsimo/vqgan-training#3]
## Issue
In `GradNormFunction`, gradient normalization is done over `dim=0`:

```python
grad_output_norm = torch.linalg.norm(grad_output, dim=0, keepdim=True)
```

`GradNormFunction` for reference:

```python
class GradNormFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):

        return x.clone()

    @staticmethod
    def backward(ctx, grad_output):

        grad_output_norm = torch.linalg.norm(grad_output, dim=0, keepdim=True)
    
        grad_output_normalized = grad_output / (grad_output_norm + 1e-8)

        return grad_output_normalized

def gradnorm(x):
    return GradNormFunction.apply(x)

# example:

x = gradnorm(x) # x.grad is now normalized to have unit norm.
```

This normalizes each feature independently, which may not be the intended behavior for gradient normalization in neural networks.

## Question
Was normalizing over `dim=0` intentional? If not, would normalizing over all dimensions except the batch dimension be more appropriate?

## Proposed Change
```python
grad_output_norm = torch.linalg.norm(grad_output, dim=grad_output.shape[1:], keepdim=True)
```

This would normalize each sample's gradient independently, potentially providing a more useful gradient norm for most neural network applications.

Could you please clarify the intended behavior and whether this change would be beneficial?

**Comment by francois-rozet** [comment]
Good catch! By the way your last line of code would crash. `dim` expects a sequence of dimension indices, not a shape. Maybe you wanted

```python
grad_output_norm = torch.linalg.norm(grad_output, dim=list(range(1, grad_output.ndim)), keepdim=True)
```

**Comment by jaivardhankapoor** [comment]
@francois-rozet Thanks for spotting the bug. 

Indeed the normalization dimension is crucial here! After some more playing around, I've concluded that the normalization over batch dim`dim=0` indeed balances the loss gradients, **but at the same time normalizes feature-specific gradient values to approximately** $\sqrt B$.

This is a **very strong** regularization, this basically means that the gradient of each pixel of the reconstruction is roughly equal when it passed through $\hat{X}$.

## tl;dr use `dim=list(range(1, grad_output.ndim))` to preserve pixel-wise differences in gradient.



I will show that norm over `dim=0` vs `dim=list(range(1, grad_output.ndim))` both normalizes the gradient across losses, but leads to drastically different distribution of magnitudes across features (elements) of $\hat{X}$. I will assume X is of the form `[B,F]`

## Math -- Gradient norm is balanced across losses in both cases (batch-wise feature normalization and feature-wise gradient normalization)

Consider a loss function $L$ and a layer output $Y$, with $\theta$ being the layer parameters.

For any normalization function $\text{norm}(\cdot)$, we have:

$$
\frac{\partial L}{\partial \theta} = \frac{\partial L}{\partial Y} \cdot \frac{\partial Y}{\partial \theta} = \text{norm}\left(\frac{\partial L}{\partial Y}\right) \cdot \frac{\partial Y}{\partial \theta}
$$

Let's denote $\frac{\partial L}{\partial Y}$ as $G\in\mathbb{R}^{B\times F}$ for simplicity.

For both normalization methods, $\|\text{norm}(G)\|_2^2 = c$, where $c$ is a constant (1 for **gradnorm over features**, and number of features for **gradnorm over batch** normalization). Both normalization can 

Now, let's calculate the squared Frobenius norm of $\frac{\partial L}{\partial \theta}$:

$$
\begin{align*}
\left\|\frac{\partial L}{\partial \theta}\right\|_F^2 &= \left\|\text{ norm}(G) \cdot \frac{\partial Y}{\partial \theta}\right\|_F^2 \\
&= \text{Tr}\left(\left(\text{norm}(G) \cdot \frac{\partial Y}{\partial \theta}\right)^T \cdot \left(\text{norm}(G) \cdot \frac{\partial Y}{\partial \theta}\right)\right)\\
&= \text{Tr}\left(\left(\frac{\partial Y}{\partial \theta}\right)^T \cdot \text{norm}(G)^T \cdot \text{norm}(G) \cdot \frac{\partial Y}{\partial \theta}\right) \\
&= \text{Tr}\left(\left(\frac{\partial Y}{\partial \theta}\right)^T \cdot \frac{\partial Y}{\partial \theta} \cdot \text{norm}(G)^T \cdot \text{norm}(G)\right) \\
&= \text{Tr}\left(\left(\frac{\partial Y}{\partial \theta}\right)^T \cdot \frac{\partial Y}{\partial \theta}\right) \cdot \|\text{norm}(G)\|_2^2 \\
&= \left\|\frac{\partial Y}{\partial \theta}\right\|_F^2 \cdot c
\end{align*}
$$

Therefore, for both normalization methods:

$$
\left\|\frac{\partial L}{\partial \theta}\right\|_F = \sqrt{c} \cdot \left\|\frac{\partial Y}{\partial \theta}\right\|_F
$$

This shows that the norm of the gradient with respect to the parameters is proportional to $$\left\|\frac{\partial Y}{\partial \theta}\right\|_F$$ for both normalization methods, with the only difference being the constant factor $$\sqrt{c}$$ (that can be set to 1 by division by norm).

For multiple losses $L_1$ and $L_2$, the ratio of their gradient norms will be:

$$
\frac{\left\|\frac{\partial L_1}{\partial \theta}\right\|_F}{\left\|\frac{\partial L_2}{\partial \theta}\right\|_F} = \frac{\sqrt{c_1} \cdot \left\|\frac{\partial Y}{\partial \theta}\right\|_F}{\sqrt{c_2} \cdot \left\|\frac{\partial Y}{\partial \theta}\right\|_F} = \sqrt{\frac{c_1}{c_2}}
$$

If the same normalization method is used for both losses, $c_1 = c_2$, and the ratio becomes 1, showing that **both losses contribute equally to the overall gradient magnitude.**

## Individual gradient elements are roughly equal for gradient normalization over batches `dim=0`, but retain different values for gradient normalization over features `dim=list(range(1, grad_output.ndim))`

This can also be shown with math, but easier to just simulate. Use the following code snippet:

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# parameters for the sweep
batch_sizes = [5, 10, 20, 50]  # different batch sizes (B)
feature_sizes = [5, 10, 20, 50]  # different feature sizes (F)
num_randomizations = 3000  # adjust the number of randomizations to balance time

# function to perform batch-wise normalization
def batch_norm(grads):
    norms = np.linalg.norm(grads, axis=0)
    normalized_grads = grads / (norms + 1e-8)
    return normalized_grads

# function to perform feature-wise normalization
def feature_norm(grads):
    norms = np.linalg.norm(grads, axis=1, keepdims=True)
    normalized_grads = grads / (norms + 1e-8)
    return normalized_grads

# dictionaries to store results for batch-wise and feature-wise normalization
batch_wise_results = {}
feature_wise_results = {}

batch_wise_variance_results = {}
feature_wise_variance_results = {}

# loop over batch sizes and feature sizes for both types of normalization
for batch_size in batch_sizes:
    for num_features in feature_sizes:
        # Batch-wise normalization (magnitude)
        avg_magnitudes_batch = []
        var_magnitudes_batch = []
        for _ in range(num_randomizations):
            grads = np.random.randn(batch_size, num_features)
            normalized_grads = batch_norm(grads)
            avg_magnitude = np.mean(np.abs(normalized_grads))
            variance_magnitude = np.var(normalized_grads)
            avg_magnitudes_batch.append(avg_magnitude)
            var_magnitudes_batch.append(variance_magnitude)
        avg_magnitude_overall_batch = np.mean(avg_magnitudes_batch)
        variance_overall_batch = np.mean(var_magnitudes_batch)
        batch_wise_results[(batch_size, num_features)] = avg_magnitude_overall_batch
        batch_wise_variance_results[(batch_size, num_features)] = variance_overall_batch

        # Feature-wise normalization (magnitude)
        avg_magnitudes_feature = []
        var_magnitudes_feature = []
        for _ in range(num_randomizations):
            grads = np.random.randn(batch_size, num_features)
            normalized_grads = feature_norm(grads)
            summed_grads = np.sum(normalized_grads, axis=0)
            avg_magnitude = np.mean(np.abs(summed_grads))
            variance_magnitude = np.var(summed_grads)
            avg_magnitudes_feature.append(avg_magnitude)
            var_magnitudes_feature.append(variance_magnitude)
        avg_magnitude_overall_feature = np.mean(avg_magnitudes_feature)
        variance_overall_feature = np.mean(var_magnitudes_feature)
        feature_wise_results[(batch_size, num_features)] = avg_magnitude_overall_feature
        feature_wise_variance_results[(batch_size, num_features)] = variance_overall_feature

# convert results to DataFrames
batchwise_df = pd.DataFrame(
    [(bs, fs, avg_mag) for (bs, fs), avg_mag in batch_wise_results.items()],
    columns=["Batch Size", "Feature Size", "Avg Magnitude"]
)

featurewise_df = pd.DataFrame(
    [(bs, fs, avg_mag) for (bs, fs), avg_mag in feature_wise_results.items()],
    columns=["Batch Size", "Feature Size", "Avg Magnitude"]
)

batchwise_var_df = pd.DataFrame(
    [(bs, fs, var_mag) for (bs, fs), var_mag in batch_wise_variance_results.items()],
    columns=["Batch Size", "Feature Size", "Variance"]
)

featurewise_var_df = pd.DataFrame(
    [(bs, fs, var_mag) for (bs, fs), var_mag in feature_wise_variance_results.items()],
    columns=["Batch Size", "Feature Size", "Variance"]
)

# plot heatmap for feature-wise normalization (magnitude)
heatmap_data_featurewise = featurewise_df.pivot("Batch Size", "Feature Size", "Avg Magnitude")

plt.figure(figsize=(10, 8))
sns.heatmap(heatmap_data_featurewise, annot=True, fmt=".3f", cmap="viridis", cbar=True)
plt.title("Heatmap of Average Magnitude (Feature-wise Gradient Norm)")
plt.show()

# plot heatmap for batch-wise normalization (magnitude)
heatmap_data_batchwise = batchwise_df.pivot("Batch Size", "Feature Size", "Avg Magnitude")

plt.figure(figsize=(10, 8))
sns.heatmap(heatmap_data_batchwise, annot=True, fmt=".3f", cmap="viridis", cbar=True)
plt.title("Heatmap of Average Magnitude (Batch-wise Gradient Norm)")
plt.show()

# plot heatmap for feature-wise variance
heatmap_data_featurewise_var = featurewise_var_df.pivot("Batch Size", "Feature Size", "Variance")

plt.figure(figsize=(10, 8))
sns.heatmap(heatmap_data_featurewise_var, annot=True, fmt=".5f", cmap="magma", cbar=True)
plt.title("Heatmap of Variance (Feature-wise Gradient Norm)")
plt.show()

# plot heatmap for batch-wise variance
heatmap_data_batchwise_var = batchwise_var_df.pivot("Batch Size", "Feature Size", "Variance")

plt.figure(figsize=(10, 8))
sns.heatmap(heatmap_data_batchwise_var, annot=True, fmt=".5f", cmap="magma", cbar=True)
plt.title("Heatmap of Variance (Batch-wise Gradient Norm)")
plt.show()
```

![image](https://github.com/user-attachments/assets/6a3ddb05-b4b8-46ae-9c8d-8b38b108767b)

The variance is MUCH lower per-feature, that can't be good! (Or maybe this is a new form of regularization that's actually beneficial?)

@cloneofsimo Is it possible to do an ablation on this?

**Comment by cloneofsimo** [comment]
Hm yes this makes a lot of sense. ive explored with two variants

```
grad_output_norm = torch.linalg.norm(grad_output)
```
and
```
grad_output_norm = torch.linalg.norm(grad_output, dim=0, keepdim=True)
```

Later performed better, while former was lot closer to original vqgan training.
its clear we have a lot more to explore!
Thank you for the discussion!

**Comment by cloneofsimo** [comment]
*note* I am not seeing any major performance differences atm, but further experiments needs to be done to back my claim. I do think using batchwise activation norm makes more sense, (at least, intuitively), so i'll keep this changes!

**Comment by cloneofsimo** [comment]
again, thanks for the discussion, I think it worth very much to explore the gradient balancing and how to do it correctly beyond just this context.

**Comment by jaivardhankapoor** [comment]
Thanks for running the ablation!
